File: core/agent/tools/table.py
"""table_execute - primary chat-agent table sink.

This is the domain-level table interface exposed to the primary thread agent.
It does NOT launch table population itself: this tool runs inside the OpenCode
tool process, while live table-population tasks and their SSE relay are owned by
the server process. The tool records the agent's table request in the chat
worktree; the server reads it after the agent turn, pins the first-class table,
launches the existing table-population worker, and streams the table element to
the thread.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from _common import ToolError, load_request, optional_string, require_string
from _run_sql import load_context

_TABLE_REQUEST_FILE = "table_request.json"


def _ensure_chat_worktree(cwd: Path) -> None:
    context = load_context(cwd)
    if context.get("mode") != "chat":
        raise ToolError(
            "table_execute can only run in a chat worktree (context mode must be "
            "`chat`)."
        )


def _default_title(request_text: str) -> str:
    title = " ".join(request_text.split())
    if len(title) <= 80:
        return title
    return title[:77].rstrip() + "..."


def _record_table_request(request: dict[str, Any], cwd: Path) -> dict[str, Any]:
    _ensure_chat_worktree(cwd)
    request_text = require_string(request, "request")
    title = optional_string(request.get("title"), "title") or _default_title(
        request_text
    )
    payload = {
        "title": title,
        "request": request_text,
        "source_template": optional_string(
            request.get("source_template"), "source_template"
        ),
        "dataset_id": optional_string(request.get("dataset_id"), "dataset_id"),
        "description": optional_string(request.get("description"), "description"),
    }
    path = cwd / _TABLE_REQUEST_FILE
    if path.exists():
        raise ToolError(
            "a table request has already been recorded for this agent turn; call "
            "table_execute at most once."
        )
    path.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
    return {"ok": True, "title": title}


def main() -> None:
    try:
        result = _record_table_request(load_request(), Path.cwd())
        print(json.dumps(result, ensure_ascii=False, indent=2))
    except (ToolError, json.JSONDecodeError) as exc:
        print(json.dumps({"ok": False, "error": str(exc)}, indent=2))
        raise SystemExit(2) from None


if __name__ == "__main__":
    main()

File: core/agent/tools/table.ts
import { tool } from "@opencode-ai/plugin"
import path from "path"

import { pythonInterpreter, toolsDir } from "./_python"

// Domain table sink. Tool name: table_execute (opencode derives
// <file>_<export>). Shells to table.py, which records the request in the chat
// worktree; the server owns pinning the table and launching population.
async function run(args: Record<string, unknown>, context: any) {
  const script = path.join(toolsDir, "table.py")
  const result = await Bun.$`${pythonInterpreter()} ${script} ${JSON.stringify(args)}`
    .cwd(context.directory ?? context.worktree)
    .nothrow()
  if (result.exitCode === 0) return result.stdout.toString()
  const stderr = result.stderr.toString().trim()
  const stdout = result.stdout.toString().trim()
  return stdout || stderr || JSON.stringify({ ok: false, error: "table exited non-zero" })
}

export const execute = tool({
  description: `Create and populate a table for the user's request. Use this when the user is asking for a structured table/audit/worklist rather than a direct natural-language answer.

You do not launch a worker or subagent yourself. Call this tool once with the domain request; the server will pin the table, start the table-population machinery, and show the table in the thread.

  - {"title":"Cord pH audit table",
     "request":"Create one row per birth with delivery mode and cord arterial pH.",
     "source_template":"cord-ph",
     "dataset_id":"ds-term-babies",
     "description":"Any concise notes that help explain the table."}

"source_template", "dataset_id", and "description" are optional. If you are unsure which template or Dataset applies, omit it or ask the user a clarification first. Returns {ok:true, title} once the table request is recorded.`,
  args: {
    request: tool.schema.string().describe("The complete table request in plain language, including requested rows/columns/scope."),
    title: tool.schema.string().optional().describe("Short table title. Omit to derive it from request."),
    source_template: tool.schema.string().optional().describe("Optional seeded audit/template id when known, e.g. \"cord-ph\"."),
    dataset_id: tool.schema.string().optional().describe("Optional Dataset id when the table should be scoped to a saved Dataset."),
    description: tool.schema.string().optional().describe("Optional concise notes for the persisted table."),
  },
  execute: run,
})
